session/filetransfer.go

```diff
@@ -0,0 +1,587 @@
+		return errors.Wrap(err, "invalid request parameters")
+	}
+
+	select {
+	case h.mutex <- struct{}{}:
+		go h.FileUploadHandler(msg, params, w) //nolint:errcheck
+	default:
+		err = errors.New("another file transfer is in progress")
+		return err
+	}
+	return nil
+}
+
+var atomicSuffix uint32
+
+func createWrOnlyTempFile(params FileInfo) (fd *os.File, err error) {
+	for i := 0; i < 100; i++ {
+		suffix := atomic.AddUint32(&atomicSuffix, 1)
+		filename := *params.Path + fmt.Sprintf(".%08x%02x", suffix, i)
+		fd, err = os.OpenFile(filename, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0200)
+		if os.IsExist(err) {
+			continue
+		} else if err != nil {
+			break
+		}
+		break
+	}
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create file")
+	}
+	return fd, nil
+}
+
+func (h *FileTransferHandler) FileUploadHandler(
+	msg *ws.ProtoMsg,
+	params FileInfo,
+
+	w ResponseWriter,
+) (err error) {
+	var (
+		fd      *os.File
+		closeFd bool
+	)
+	defer func() {
+		if fd != nil {
+			if closeFd {
+				errClose := fd.Close()
+				if errClose != nil {
+					log.Warnf("error closing file: %s", errClose.Error())
+				}
+			}
+			errRm := os.Remove(fd.Name())
+			if errRm != nil {
+				log.Errorf(
+					"error removing file after aborting upload: %s",
+					errRm.Error(),
+				)
+			}
+		}
+		if err != nil {
+			log.Error(err.Error())
+			if errors.Cause(err) != errFileTransferAbort {
+				h.Error(msg, w, err)
+			}
+		}
+		<-h.mutex
+	}()
+
+	fd, err = createWrOnlyTempFile(params)
+	if err != nil {
+		h.Error(msg, w, errors.Wrap(err, "failed to create target file"))
+		return err
+	}
+	closeFd = true
+
+	err = w.WriteProtoMsg(&ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:      ws.ProtoTypeFileTransfer,
+			MsgType:    wsft.MessageTypeACK,
+			SessionID:  msg.Header.SessionID,
+			Properties: map[string]interface{}{"offset": int64(0)},
+		},
+	})
+	if err != nil {
+		log.Errorf("failed to respond to client: %s", err.Error())
+		return errFileTransferAbort
+	}
+
+	_, err = h.writeFile(w, fd)
+	if err != nil {
+		return err
+	}
+	// Set the final permissions and owner.
+	err = fd.Chmod(os.FileMode(*params.Mode) & os.ModePerm)
+	if err != nil {
+		return errors.Wrap(err, "failed to set file permissions")
+	}
+	err = fd.Chown(int(*params.UID), int(*params.GID))
+	if err != nil {
+		return errors.Wrap(err, "failed to set file owner")
+	}
+
+	closeFd = false
+	filename := fd.Name()
+	errClose := fd.Close()
+	if errClose != nil {
+		log.Warnf("error closing file: %s", errClose.Error())
+	}
+	err = os.Rename(filename, *params.Path)
+	if err != nil {
+		return errors.Wrap(err, "failed to commit uploaded file")
+	}
+	fd = nil
+	return err
+}
+
+func (h *FileTransferHandler) writeFile(w ResponseWriter, dst *os.File) (int64, error) {
+	var (
+		done   bool
+		open   bool
+		err    error
+		i      int
+		offset int64
+		msg    *ws.ProtoMsg
+	)
+	// Convenience clojure for decoding file chunk and writing to destination file.
```

Comment:
is "clojure" an intentional reference to Lisp?

Reply:
Oops, that should say "function closure" :see_no_evil: